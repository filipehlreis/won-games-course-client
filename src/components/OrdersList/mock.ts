export default [
  {
    id: '1',
    paymentInfo: {
      flag: 'mastercard',
      img: '/img/mastercard.png',
      number: '**** **** **** 4326',
      purchaseDate: 'Purchase made on 07/20/2020 at 20:32',
    },
    games: [
      {
        id: '1',
        img: 'https://source.unsplash.com/user/willianjusten/151x71',
        title: 'Red Dead Redemption 2',
        price: 'R$ 215,00',
        downloadLink:
          'https://wongames.com/game/download/yuYT56Tgh431LkjhNBgdf',
      },
    ],
  },
  {
    id: '2',
    paymentInfo: {
      flag: 'visa',
      img: '/img/visa.png',
      number: '**** **** **** 5423',
      purchaseDate: 'Purchase made on 08/20/2020 at 21:12',
    },
    games: [
      {
        id: '1',
        img: 'https://source.unsplash.com/user/willianjusten/151x71',
        title: 'Red Dead Redemption 2',
        price: 'R$ 215,00',
        downloadLink: 'https://wongames.com/game/download/kjhejl867asd76DEh',
      },
    ],
  },
];
