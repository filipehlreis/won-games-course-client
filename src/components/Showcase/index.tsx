import * as S from './styles';

export const Showcase = () => {
  return (
    <S.Wrapper>
      <h1>Showcase</h1>
    </S.Wrapper>
  );
};
