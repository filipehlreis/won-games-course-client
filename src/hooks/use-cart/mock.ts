import { QUERY_GAMES } from 'graphql/queries/games';

export const gamesMock = {
  request: {
    query: QUERY_GAMES,
    variables: { filters: { id: { in: ['1', '2'] } } },
  },
  result: {
    data: {
      games: {
        data: [
          {
            id: '1',
            attributes: {
              name: 'Sample Game 1',
              slug: 'sample_game_1',
              short_description: 'Sample description 1',
              price: 9.9,
              developers: {
                data: [
                  {
                    attributes: {
                      name: 'sample developer 1',
                    },
                  },
                ],
              },
              cover: {
                data: {
                  attributes: {
                    url: '/sample_game_1.jpg',
                  },
                },
              },
              __typename: 'Game',
            },
          },
          {
            id: '2',
            attributes: {
              name: 'Sample Game 2',
              slug: 'sample_game_2',
              short_description: 'Sample description 2',
              price: 10.5,
              developers: {
                data: [
                  {
                    attributes: {
                      name: 'sample developer 2',
                    },
                  },
                ],
              },
              cover: {
                data: {
                  attributes: {
                    url: '/sample_game_2.jpg',
                  },
                },
              },
              __typename: 'Game',
            },
          },
        ],
      },
    },
  },
};

export const cartItems = [
  {
    id: '1',
    img: 'http://localhost:1337/sample_game_1.jpg',
    price: '$9.90',
    title: 'Sample Game 1',
  },
  {
    id: '2',
    img: 'http://localhost:1337/sample_game_2.jpg',
    price: '$10.50',
    title: 'Sample Game 2',
  },
];
