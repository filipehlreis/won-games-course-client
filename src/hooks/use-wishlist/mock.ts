import {
  MUTATION_CREATE_WISHLIST,
  MUTATION_UPDATE_WISHLIST,
} from 'graphql/mutations/wishlist';
import { QUERY_WISHLIST } from 'graphql/queries/wishlist';

const gameMock = (id: string) => ({
  __typename: 'GameEntity',
  id,
  attributes: {
    name: `Sample Game ${id}`,
    slug: `sample-game-${id}`,
    cover: {
      data: {
        attributes: {
          url: '/sample-game.jpg',
        },
      },
    },
    developers: {
      data: [
        {
          attributes: {
            name: 'sample developer',
          },
        },
      ],
    },

    price: 10.5,
  },
});

export const wishlistMock = {
  request: {
    query: QUERY_WISHLIST,
    context: { session: { jwt: '123' } },
    variables: {
      identifier: 'lorem@ipsum.com',
    },
  },
  result: {
    data: {
      wishlists: {
        data: [
          {
            id: 1,
            attributes: {
              games: {
                __typename: 'GameRelationResponseCollection',
                data: [gameMock('1'), gameMock('2')],
              },
            },
          },
        ],
      },
    },
  },
};

export const createWishlistMock = {
  request: {
    query: MUTATION_CREATE_WISHLIST,
    context: { session: { jwt: '123' } },
    variables: {
      input: {
        games: ['3'],
        user: 1,
      },
    },
  },
  result: {
    data: {
      createWishlist: {
        data: {
          id: '1',
          attributes: {
            games: {
              data: [gameMock('3')],
            },
          },
        },
      },
    },
  },
};

export const updateWishlistMock = {
  request: {
    query: MUTATION_UPDATE_WISHLIST,
    context: { session: { jwt: '123' } },
    variables: {
      id: 1,
      data: {
        games: ['1', '2', '3'],
        user: 1,
      },
    },
  },
  result: {
    data: {
      updateWishlist: {
        data: {
          id: '1',
          attributes: {
            games: {
              data: [gameMock('1'), gameMock('2'), gameMock('3')],
            },
          },
        },
      },
    },
  },
};

export const wishlistItems = [
  {
    id: '1',
    title: 'Sample Game 1',
    slug: 'sample-game-1',
    developer: 'sample developer',
    img: 'http://localhost:1337/sample-game.jpg',
    price: 10.5,
  },
  {
    id: '2',
    title: 'Sample Game 2',
    slug: 'sample-game-2',
    developer: 'sample developer',
    img: 'http://localhost:1337/sample-game.jpg',
    price: 10.5,
  },
  {
    id: '3',
    title: 'Sample Game 3',
    slug: 'sample-game-3',
    developer: 'sample developer',
    img: 'http://localhost:1337/sample-game.jpg',
    price: 10.5,
  },
];
