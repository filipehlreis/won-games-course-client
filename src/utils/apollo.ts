import {
  ApolloClient,
  HttpLink,
  InMemoryCache,
  NormalizedCacheObject,
} from '@apollo/client';
// import { concatPagination } from '@apollo/client/utilities';
import { useMemo } from 'react';
// import { QueryGames_games, QueryGames } from '../graphql/generated/QueryGames';

let apolloClient: ApolloClient<NormalizedCacheObject | null>;

function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: new HttpLink({ uri: 'http://localhost:1337/graphql' }),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            games: {
              keyArgs: ['sort', 'filters'],
              merge: true,
            },
          },
        },
      },
    }),
  });
}

// function createApolloClient() {
//   return new ApolloClient({
//     ssrMode: typeof window === 'undefined',
//     link: new HttpLink({ uri: 'http://localhost:1337/graphql' }),
//     cache: new InMemoryCache({
//       typePolicies: {
//         Query: {
//           fields: {
//             games: concatPagination(['where', 'sort']),
//           },
//         },
//       },
//     }),
//   });
// }

// function createApolloClient() {
//   return new ApolloClient({
//     ssrMode: typeof window === 'undefined',
//     link: new HttpLink({ uri: 'http://localhost:1337/graphql' }),
//     cache: new InMemoryCache({
//       typePolicies: {
//         Query: {
//           fields: {
//             games: {
//               // Don't cache separate results based on
//               // any of this field's arguments.
//               keyArgs: false,

//               // Concatenate the incoming list items with
//               // the existing list items.
//               merge(existing = [], incoming: QueryGames_games) {
//                 // console.log('existing', existing);
//                 console.log('incoming', incoming.data);
//                 // data?.games?.data

//                 return [...existing, ...incoming.data];
//                 // return [...existing, ...incoming.data];
//               },
//             },
//           },
//         },
//       },
//     }),
//   });
// }

export function initializeApollo(initialState = {}) {
  // serve para verificar se ja existe uma isntancia ,para nao criar outra
  const apolloClientGlobal = apolloClient ?? createApolloClient();

  // recuperando os dados de cache
  if (Object.keys(initialState).length) {
    apolloClientGlobal.cache.restore(initialState);
  }

  // sempre inicializando no SSR com cache limpo
  if (typeof window === 'undefined') return apolloClientGlobal;

  apolloClient = apolloClient ?? apolloClientGlobal;

  return apolloClient;
}

export function useApollo(initialState = {}) {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
