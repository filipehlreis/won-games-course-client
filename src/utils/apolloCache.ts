import { InMemoryCache } from '@apollo/client';

export default new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        games: {
          keyArgs: ['sort', 'filters'],
          merge: true,
        },
      },
    },
  },
});
