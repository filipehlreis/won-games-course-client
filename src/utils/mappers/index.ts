import { QueryGames_games } from '../../graphql/generated/QueryGames';
import {
  QueryHome_banners,
  QueryHome_sections_data_attributes_freeGames_highlight,
  QueryHome_sections_data_attributes_popularGames_games,
} from '../../graphql/generated/QueryHome';

export const bannerMapper = (banners: QueryHome_banners) => {
  return banners!.data.map((banner) => ({
    img: `http://localhost:1337${
      banner.attributes!.image.data?.attributes?.url
    }`,
    title: banner.attributes!.title,
    subtitle: banner.attributes!.subtitle,
    buttonLabel: banner.attributes!.button?.label,
    buttonLink: banner.attributes!.button?.link,
    ...(banner.attributes!.ribbon && {
      ribbon: banner.attributes!.ribbon?.text,
      ribbonColor: banner.attributes!.ribbon?.color,
      ribbonSize: banner.attributes!.ribbon?.size,
    }),
  }));
};

export const gamesMapper = (
  games:
    | QueryGames_games
    | QueryHome_sections_data_attributes_popularGames_games
    | null
    | undefined,
) => {
  return games
    ? games.data.map((game) => ({
        id: game.id,
        title: game.attributes!.name,
        slug: game.attributes!.slug,
        developer: game.attributes!.developers!.data[0].attributes!.name,
        img: `http://localhost:1337${
          game.attributes!.cover?.data!.attributes!.url
        }`,
        price: game.attributes!.price,
      }))
    : [];
};

export const highlightMapper = (
  highlight:
    | QueryHome_sections_data_attributes_freeGames_highlight
    | null
    | undefined,
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `http://localhost:1337${highlight.background.data?.attributes?.url}`,
        floatImage: `http://localhost:1337${highlight.floatImage?.data?.attributes?.url}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment,
      }
    : {};
};
