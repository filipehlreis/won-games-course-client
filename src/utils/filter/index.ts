/* eslint-disable prettier/prettier */
import { ItemProps } from 'components/ExploreSidebar';
import { GameFiltersInput } from 'graphql/generated/globalTypes';
import { ParsedUrlQueryInput } from 'querystring';

type ParseArgs = {
  queryString: ParsedUrlQueryInput;
  filterItems: Pick<ItemProps, 'type' | 'name'>[];
};



/**
 *
 *
 * TODO: It's necessary to create a Sort method to fix the duplicated results
 * when fetching more games.
 *
 *
 *
 *
 */







/**
 *
 * queryString
 * price_lte=0&platforms=windows&platforms=linux&categories=sports&sort=price%3Aasc

 para

 where

      price: { lte: 600 },
      and: [
        {
          platforms: {
            name: {
              containsi: 'mac',
            },
          },
        },
        {
          platforms: {
            name: {
              containsi: 'windows',
            },
          },
        },
        {
          platforms: {
            name: {
              containsi: 'linux',
            },
          },
        },
      ],

 */

// const queryString = {
//   price_lte: 100,
//   platforms: ['mac', 'windows', 'linux'],
//   sort: 'price:asc',
// };

export const parseQueryStringToWhereNew = ({
  queryString,
  // filterItems,
}: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: GameFiltersInput = {};

  // console.log('queryString dentro da funcao', queryString);

  Object.keys(queryString)
    .filter((item) => item !== 'sort')
    .forEach((key) => {

      // console.log('entrei na funcao parseQueryStringToWhereNew()')
      // const item = filterItems?.find((item) => item.name === key);
      // const isCheckbox = item?.type === 'checkbox';

      // console.log('isCheckbox', isCheckbox, 'item', item);
      // console.log('key', key);
      // console.log('item da funcao', item);
      // console.log('queryString[key] da funcao', queryString[key]);
      // const queryStringVetor = Array.isArray(queryString[key]) ? queryString[key] : [queryString[key]]
      // ? Object({ and: queryString[key] })
      // obj['and'] = !isCheckbox
      //   ? queryString[key]
      //   : [Object({ [key]: Object({ containsi: queryString[key] }) })];

      key === 'price' && (obj.price = Object({ lte: Number(queryString[key]) }));
      // key === 'platforms' && (

      //   obj.and = Object({ platforms: { containsi: queryString[key] } })
      // );

      const queryStringVetor = Array.isArray(queryString[key]) ? queryString[key] : [queryString[key]]

      key === 'categories' && (
        <[]>queryStringVetor).map((value: string) => {
          obj.and?.length
            ? obj.and?.push({ categories: { name: { containsi: value } } })
            : obj.and = [{ categories: { name: { containsi: value } } }]
        }
        );

      key === 'platforms' && (
        <[]>queryStringVetor).map((value: string) => {
          obj.and?.length
            ? obj.and?.push({ platforms: { name: { containsi: value } } })
            : obj.and = [{ platforms: { name: { containsi: value } } }]
        }
        );

      // !isCheckbox
      //   ? (obj[key] =
      //     key === 'price'
      //       ? Object({ lte: queryString[key] })
      //       : queryString[key])
      //   : (
      //     queryString[key]?.map((valor: string) => {
      //       obj.and = [Object({ [key]: Object({ containsi: valor }) }),]
      //     })
      //   );

      // obj[key] = !isCheckbox
      //   ? key === 'price'
      //     ? Object({ lte: queryString[key] })
      //     : queryString[key]
      //   : [Object({ [key]: Object({ containsi: queryString[key] }) })];

      // console.log('obj[key] da funcao', obj[key]);
    });

  // console.log('obj da funcao', JSON.stringify(obj, null, 4));

  // console.log('\n\nimprimindo objeto', JSON.stringify(obj, null, 4))
  return obj;
};

// export const parseQueryStringToWhere = ({
//   queryString,
//   filterItems,
// }: ParseArgs) => {
//   // eslint-disable-next-line @typescript-eslint/no-explicit-any
//   const obj: any = {};
//   // console.log('queryString', queryString);

//   Object.keys(queryString)
//     .filter((item) => item !== 'sort')
//     .forEach((key) => {
//       const item = filterItems?.find((item) => item.name === key);
//       const isCheckbox = item?.type === 'checkbox';

//       obj[key] = !isCheckbox
//         ? queryString[key]
//         : { name_contains: queryString[key] };
//     });

//   // console.log('obj', obj);

//   return obj;
// };

export const parseQueryStringToFilter = ({
  queryString,
  filterItems,
}: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {};

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((item) => item.name === key);
    const isCheckbox = item?.type === 'checkbox';
    const isArray = Array.isArray(queryString[key]);

    obj[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key];
  });

  return obj;
};

/**
 *
 *
 *
 *
 *
 *
 */

// const objeto = {
//   limit: 200,
//   sort: 'price:desc',
//   filters: {
//     price: { lte: 600 },
//     and: [
//       {
//         platforms: {
//           name: {
//             containsi: 'mac',
//           },
//         },
//       },
//       {
//         platforms: {
//           name: {
//             containsi: 'windows',
//           },
//         },
//       },
//       {
//         platforms: {
//           name: {
//             containsi: 'linux',
//           },
//         },
//       },
//       {
//         categories: {
//           name: {
//             containsi: 'adventure',
//           },
//         },
//       },
//       {
//         categories: {
//           name: {
//             containsi: 'fantasy',
//           },
//         },
//       },
//       {
//         categories: {
//           name: {
//             containsi: 'action',
//           },
//         },
//       },
//     ],
//   },
// };

/**
 *
 *
 *
 *
 *
 */

// import { ItemProps } from 'components/ExploreSidebar';
// import { ParsedUrlQueryInput } from 'querystring';

// type ParseArgs = {
//   queryString: ParsedUrlQueryInput;
//   filterItems: Pick<ItemProps, 'type' | 'name'>[];
// };

// export const parseQueryStringToWhere = ({
//   queryString,
//   filterItems,
// }: ParseArgs) => {
//   // eslint-disable-next-line @typescript-eslint/no-explicit-any
//   const obj: any = {};
//   console.log('queryString', queryString);

//   Object.keys(queryString)
//     .filter((item) => item !== 'sort')
//     .forEach((key) => {
//       const item = filterItems?.find((item) => item.name === key);
//       const isCheckbox = item?.type === 'checkbox';

//       obj[key] = !isCheckbox
//         ? queryString[key]
//         : { name_contains: queryString[key] };
//     });

//   console.log('obj', obj);

//   return obj;
// };

// export const parseQueryStringToFilter = ({
//   queryString,
//   filterItems,
// }: ParseArgs) => {
//   // eslint-disable-next-line @typescript-eslint/no-explicit-any
//   const obj: any = {};

//   Object.keys(queryString).forEach((key) => {
//     const item = filterItems?.find((item) => item.name === key);
//     const isCheckbox = item?.type === 'checkbox';
//     const isArray = Array.isArray(queryString[key]);

//     obj[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key];
//   });
//   return obj;
// };
