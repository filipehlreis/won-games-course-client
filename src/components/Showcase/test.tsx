import 'match-media-mock';
import 'session.mock';

import { screen, render } from 'utils/test-utils';
import highlightMock from 'components/Highlight/mock';
import gamesMock from 'components/GameCardSlider/mock';

import Showcase from '.';

const props = {
  title: 'Most Popular',
  highlight: highlightMock,
  games: gamesMock.slice(0, 1),
};

describe('<Showcase />', () => {
  it('should render the heading', () => {
    render(<Showcase {...props} />);

    expect(
      screen.getByRole('heading', { name: /most popular/i }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole('heading', { name: highlightMock.title }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole('heading', { name: gamesMock[0].title }),
    ).toBeInTheDocument();
  });

  it('should render without title', () => {
    render(<Showcase games={props.games} highlight={props.highlight} />);

    screen.getByRole('heading', { name: highlightMock.title });
    screen.getByRole('heading', { name: gamesMock[0].title });

    expect(
      screen.queryByRole('heading', { name: /most popular/i }),
    ).not.toBeInTheDocument();
  });

  it('should render without highlight', () => {
    render(<Showcase title={props.title} games={props.games} />);

    screen.getByRole('heading', { name: /most popular/i });
    screen.getByRole('heading', { name: gamesMock[0].title });

    expect(
      screen.queryByRole('heading', { name: highlightMock.title }),
    ).not.toBeInTheDocument();
  });

  it('should render without games', () => {
    render(<Showcase title={props.title} highlight={props.highlight} />);

    screen.getByRole('heading', { name: /most popular/i });
    screen.getByRole('heading', { name: highlightMock.title });

    expect(
      screen.queryByRole('heading', { name: gamesMock[0].title }),
    ).not.toBeInTheDocument();
  });
});
