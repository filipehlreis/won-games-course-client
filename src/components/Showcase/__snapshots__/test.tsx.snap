// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Showcase /> should render the heading 1`] = `
<main
  class=""
>
  <h1>
    Showcase
  </h1>
</main>
`;
